languages/rust/src/lib.rs

```diff
@@ -3,214 +3,38 @@
-            ModuleFFI::User(ref user) => &user.module_meta,
-        }
-    }
-}
-
-/// Standard Modules are modules that are built into Flipper and always
-/// available on the device.
-///
-/// Standard Modules must implement `new` and `bind`. The `new` constructor
-/// gives an instance of the module that is attached to the "default"
-/// Flipper. `bind` gives an instance which is attached to a specific
-/// Flipper.
-pub trait StandardModule {
-    fn new() -> Self;
-    fn bind(flipper: &Flipper) -> Self;
-}
-
-/// User Modules are modules which are written in C by the user.
-///
-/// When a user creates a module, they give it a name which is used by
-/// Flipper to load and bind to it. To use it from the Rust bindings,
-/// the user must specify the name so that rust can find the module.
-pub trait UserModule<'a>: From<UserModuleFFI> {
-    const NAME: &'a str;
-    fn new() -> Self;
-
-    /// Binds an instance of a User Module to the given Flipper.
-    ///
-    /// ```
-    /// use flipper::{Flipper, UserModule, ModuleFFI, UserModuleFFI};
-    ///
-    /// struct MyModule {
-    ///     ffi: ModuleFFI,
-    /// }
-    ///
-    /// impl<'a> UserModule<'a> for MyModule {
-    ///     const NAME: &'a str = "My module";
-    ///     fn new() -> Self {
-    ///         MyModule {
-    ///             ffi: ModuleFFI::User(UserModuleFFI::uninitialized(Self::NAME)),
-    ///         }
-    ///     }
-    /// }
-    ///
-    /// impl From<UserModuleFFI> for MyModule {
-    ///     fn from(module: UserModuleFFI) -> Self {
-    ///         MyModule {
-    ///             ffi: ModuleFFI::User(module),
-    ///         }
-    ///     }
-    /// }
-    ///
-    /// impl MyModule {
-    ///     fn myFunc(&self) {
-    ///         // Do FMR invocation
-    ///     }
-    /// }
-    ///
-    /// let flipper = Flipper::attach();
-    ///
-    /// // Any of the following will bind "MyModule"
-    /// let myModule = MyModule::new();          // Attaches to default ("active") flipper
-    /// let myModule = MyModule::bind(&flipper); // Attaches to specified flipper
-    ///
-    /// myModule.myFunc();
-    /// ```
-    fn bind(flipper: &Flipper) -> Self {
-        let mut module = UserModuleFFI::uninitialized(Self::NAME);
-        unsafe { lf_bind(&mut module.module_meta, flipper.device); }
-        Self::from(module)
-    }
-}
-
 #[link(name = "flipper")]
 extern {
     fn flipper_attach() -> _lf_device;
     fn carbon_attach_hostname(hostname: *const c_char) -> _lf_device;
-    fn lf_bind(module: *mut _lf_module, device: *const c_void) -> c_int;
 }
 
-pub struct Flipper {
-    /// A reference to an active Flipper profile in libflipper. This
-    /// is used when communicating with libflipper to specify which
-    /// device functions should be executed on.
-    device: _lf_device,
-}
+
+pub struct Flipper;
 
 impl Flipper {
-    pub fn attach() -> Self {
+    pub fn attach() -> _lf_device {
```

Comment:
The `_lf_device` type shouldn't appear in the public API at all. If it needs to exist, it should be stored as a private field in a struct.

Reply:
👍